src/main/java/com/powsybl/openloadflow/network/impl/PropagatedContingency.java

```diff
@@ -0,0 +1,332 @@
+            double nominalV = shunt.getTerminal().getVoltageLevel().getNominalV();
+            shuntIdsToShift.put(shunt.getId(), shunt.getB() * nominalV * nominalV / PerUnit.SB);
+        }
+
+        return new PropagatedContingency(contingency, index, branchIdsToOpen, hvdcIdsToOpen, switchesToOpen,
+                                         generatorIdsToLose, loadIdsToShift, shuntIdsToShift);
+    }
+
+    private static boolean isCoupler(Switch s) {
+        VoltageLevel.NodeBreakerView nbv = s.getVoltageLevel().getNodeBreakerView();
+        Terminal terminal1 = nbv.getTerminal1(s.getId());
+        Terminal terminal2 = nbv.getTerminal2(s.getId());
+        if (terminal1 == null || terminal2 == null) {
+            return false; // FIXME: this can be a coupler, a traverser could be used to detect it
+        }
+        Connectable<?> c1 = terminal1.getConnectable();
+        Connectable<?> c2 = terminal2.getConnectable();
+        return c1 != c2 && c1.getType() == IdentifiableType.BUSBAR_SECTION && c2.getType() == IdentifiableType.BUSBAR_SECTION;
+    }
+
+    public Optional<LfContingency> toLfContingency(LfNetwork network, GraphDecrementalConnectivity<LfBus> connectivity,
+                                                   boolean useSmallComponents) {
+        // find contingency branches that are part of this network
+        Set<LfBranch> branches = new HashSet<>(1);
+        for (String branchId : branchIdsToOpen) {
+            LfBranch branch = network.getBranchById(branchId);
+            if (branch != null) {
+                branches.add(branch);
+            }
+        }
+
+        // check if contingency split this network into multiple components
+        if (branches.isEmpty()
+                && shuntIdsToShift.isEmpty()
+                && loadIdsToShift.isEmpty()
+                && generatorIdsToLose.isEmpty()) {
+            return Optional.empty();
+        }
+
+        // update connectivity with triggered branches
+        for (LfBranch branch : branches) {
+            connectivity.cut(branch.getBus1(), branch.getBus2());
+        }
+
+        // add to contingency description buses and branches that won't be part of the main connected
+        // component in post contingency state
+        Set<LfBus> buses;
+        if (useSmallComponents) {
+            buses = connectivity.getSmallComponents().stream().flatMap(Set::stream).collect(Collectors.toSet());
+        } else {
+            int slackBusComponent = connectivity.getComponentNumber(network.getSlackBus());
+            buses = network.getBuses().stream().filter(b -> connectivity.getComponentNumber(b) != slackBusComponent).collect(Collectors.toSet());
+        }
+        buses.forEach(b -> branches.addAll(b.getBranches()));
+
+        Map<LfShunt, Double> shunts = new HashMap<>(1);
+        for (var e : shuntIdsToShift.entrySet()) {
+            // FIXME does not work when multiple shunts connected to a bus because IDs won't be found !!!!
```

Comment:
This was not the purpose of this PR, but shunt contingency during SA cannot work as we do not search with the right ID since we have aggregated shunts by buses with a deviated ID. To fix later?

Reply:
I don't understand the problem. For each IIDM shunt id (`shunt.getIds()` gives a list of IIDM shunt ids) we can get the aggregated shunt. So for me if we loose 2 shunts connected to the same bus, the same aggregated LfShunt will be shift twice. The issue is maybe to use a HashMap or not to check if the LfShunt is already in the map ?

Reply:
Right, I removed my comment